### DISCUSSION TITLE
Custom Trriger Type

### DISCUSSION THREAD

**Discussion by danielpalacian** [quartznet/quartznet#1521] [Q&A]
Is it possible to create a custom trigger and use it after in the Scheduler? 

I would want to Create a custom Trigger that Inherits the CronTrigger and to override the GetNextFireTimeUtc() logic for example. 

Could this be possible?

**Comment by lahma** [comment]
Yes it should be possible, but there are caveats. I wouldn't use this as persisted trigger (database) as wrong type would probably be materialized, or you need to build own persistence - haven't played with it. As long as you are doing in-memory only, I *think* it should work. 

But if there's a bug or something in CronTrigger, maybe we should fix it.

**Comment by danielpalacian** [comment]
The Persistence it is at DB level and indeed it does not work as i would expect. It works with the first execution but after my Triger is not used anymore. 

I have a more specific use case. 
I have a defined Job and Trigger that i need to "register" multiple times (multiple customers). The trigger it is a Cron one (that i can't control). 
I would want to "Schedule" each one with a small "offset"(10-15seconds) . This is why i wanted to override the "CronTrriger" with a custom one that "adds" the offset when the next execution it is calculated. 
It works as expected when i first register the triggers, but next execution does not, because it does not use my custom code anymore. 

I am not keen on this solution. 
But i would want to achieve it somehow. 

Will post bellow some of my code. 


```csharp
    public class CronTriggerWithOffset : CronTriggerImpl
    {
        private readonly int offset;

        public CronTriggerWithOffset(CronTriggerImpl trigger, int offset) : base(trigger.Name, trigger.Group, trigger.JobName, trigger.JobGroup, trigger.CronExpressionString)
        {
            this.offset = offset;
        }
       
        public override DateTimeOffset? GetNextFireTimeUtc()
        {
            var date = base.GetNextFireTimeUtc();
            if (date.HasValue)
                return date.Value.AddSeconds(offset);

            return null;
        }
    }
```

**Reply by lahma** [reply]
Why do you need the offset? If it's just to reduce pressure from concurrent fires you can achieve that by limiting scheduler thread pool size (concurrent worker count) or by referencing the same job and marking it with `[DisallowConcurrentExecution]`, or even scheduler per customer.

**Reply by danielpalacian** [reply]
Will try the DisallowConcurrentExecution sounds like something that could do what i need. Even if i would want to have a little more control. 
My next q is .. would it be possible to add an "offset" to a CronExpression. So let's say i have the cron expression set ( can't control that) and i would want to add an offset of 3 seconds to it. Would it be possible ( i was trying to reverse engineer CronExpression class right now).

**Reply by lahma** [reply]
Offset doesn't feel like natural concept in cron scheduling, the expression states the exact schedule. You describe that you need an offset, but I still don't understand why.

**Reply by danielpalacian** [reply]
We need it for controlling the pressure that we put on a specific job let's say. 
We have a scheduler that controls around 100 jobs , each job having around 200 triggers ( one per customer).
We already have a number that controls (concurrent worker threads ) but that is for all the jobs. I would want to have an option to control a specific job (with the triggers) the number of jobs that i can schedule in parallel. Ideally it should be with the help of CronExpressions, But because this is something that i would want to have "generic" i can't / i would avoid parsing CronExpressions and this is why i have tried with an Offset approach. 
If i could use for example the CronExpression class and with the help of it to generate an other CronExpression, that would be even better, because i would have the option to schedule the triggers how i want after. 
Do you know an option of doing that?

**Reply by lahma** [reply]
I think there are some problems in this approach. 

First, you cron expressions are invalid in the "technical business sense" that you don't want them to be the way they were defined. Somebody might even rely on them running at defined time, not even seconds later/before (a bit hypothetical but could be source of timing issues) 

Secondly, say the offset is 3 seconds. Jobs start to take 3 seconds more than they used to (more data, temporary slowdown) you are again back to original problem with pressure. So offset won't tell you how many jobs you will be having concurrently - running with same cron expression this could be from 1-200 based on some unpredictability.

Some tools for this kind of scenario (in no particular order):

* scheduler thread count (like you already have)
* multiple schedulers (dedicated to customers, but can become cumbersome if in the hundreds, but allow configuration like one thread per customer allowed)
* custom job run shell, but will only make sense in single node case
* listeners, add variance to dispatch process (this will still have the second issue I described)
* DisallowConcurrentExecution, locks single job key, might cause waits to other customers (when sharing the scheduler)
* job logic acquire some mutex or green flag to continue with this specific job

So I don't have direct answers, I'm just afraid that offsets would introduce more unpredictability and harder to understand because cron expressions don't match the actual schedules.

**Reply by danielpalacian** [reply]
We control the "Triggers" so if it an "offset" is not a pb , not the final customer or anybody else
The job that we would want to control in this scenario, call's an external API that struggles if we do to many call's at the same time. 
So this is why i would want to "create" on the fly "CronExpressions" with some offset between them to have more control. 

 I have seen that you said about listeners. 
 Can you pls detail that approach? How could i achieve it with that? . 
 We already have Listeners to log some general stuff , but i am not aware how i could use them to achieve what i want.

**Reply by danielpalacian** [reply]
It will be a cron expression in the end. Let's forget the "offset" word. 
Is it possible to create dynamically with the help of CronExpression class other CronExpressions? 
If we have a CronExpression like "0 15 7 ? * *"  .. can i create a CronExpression like "0 16 7 ? * *" or "0 20 7 ? * *" dynamically with the CronExpression class ?

**Reply by lahma** [reply]
You could probably extract parsing code from Quartz of some other library like [cron expression descriptor](https://www.nuget.org/packages/CronExpressionDescriptor/).

I would probably just make sure that jobs are properly queued using triggers that can handle misfires and then having job with disallow concurrent execution per API endpoint. So high-level design would be job key per API endpoint and triggers having job data map having the invocation parameters which job would receive and then process based on that (jobs can get their parameters from triggers too).
